cstore_compression.c

```diff
@@ -0,0 +1,165 @@
+/*-------------------------------------------------------------------------
+ *
+ * cstore_compression.c
+ *
+ * This file contains compression/decompression functions definitions
+ * used in cstore_fdw.
+ *
+ * Copyright (c) 2016, Citus Data, Inc.
+ *
+ * $Id$
+ *
+ *-------------------------------------------------------------------------
+ */
+#include "postgres.h"
+#include "cstore_fdw.h"
+
+#if PG_VERSION_NUM >= 90500
+#include "common/pg_lzcompress.h"
+#else
+#include "utils/pg_lzcompress.h"
+#endif
+
+
+
+
+#if PG_VERSION_NUM >= 90500
+/*
+ *	The information at the start of the compressed data. This decription is taken
+ *	from pg_lzcompress in pre-9.5 version of PostgreSQL.
+ */
+typedef struct CStoreCompressHeader
+{
+	int32		vl_len_;		/* varlena header (do not touch directly!) */
+	int32		rawsize;
+} CStoreCompressHeader;
+
+/*
+ * Utilities for manipulation of header information for compressed data
+ */
+
+#define CSTORE_COMPRESS_HDRSZ		((int32) sizeof(CStoreCompressHeader))
+#define CSTORE_COMPRESS_RAWSIZE(ptr) (((CStoreCompressHeader *) (ptr))->rawsize)
+#define CSTORE_COMPRESS_RAWDATA(ptr) (((char *) (ptr)) + CSTORE_COMPRESS_HDRSZ)
+#define CSTORE_COMPRESS_SET_RAWSIZE(ptr, len) (((CStoreCompressHeader *) (ptr))->rawsize = (len))
+
+#else
+
+#define CSTORE_COMPRESS_HDRSZ		(0)
+#define CSTORE_COMPRESS_RAWSIZE(ptr) (PGLZ_RAW_SIZE((PGLZ_Header *) buffer->data))
+#define CSTORE_COMPRESS_RAWDATA(ptr) (((PGLZ_Header *) (ptr)))
+#define CSTORE_COMPRESS_SET_RAWSIZE(ptr, len) (((CStoreCompressHeader *) (ptr))->rawsize = (len))
+
+#endif
+
+
+
+/*
+ * CompressBuffer compresses the given buffer with the given compression type
+ * outputBuffer enlarged to contain compressed data. The function returns true
+ * if compression is done, returns false if compression is not done.
+ * outputBuffer is valid only if the function returns true.
+ */
+bool
+CompressBuffer(StringInfo inputBuffer, StringInfo outputBuffer,
+			   CompressionType compressionType)
+{
+	uint64 maximumLength = PGLZ_MAX_OUTPUT(inputBuffer->len) + CSTORE_COMPRESS_HDRSZ;
+	bool compressionResult = false;
+
+	if (compressionType != COMPRESSION_PG_LZ)
+	{
+		return false;
+	}
+
+	resetStringInfo(outputBuffer);
+	enlargeStringInfo(outputBuffer, maximumLength);
+
+#if PG_VERSION_NUM >= 90500
+	int32 compressedByteCount = 0;
+
+	compressedByteCount = pglz_compress((const char *) inputBuffer->data,
```

Comment:
Ok, I read&understand the API changes. But, I have a question for our implementation. What happens when `pglz_compress ` returns `0`. Is there a possibility that `outputBuffer->data` is `NULL` etc so that cstore crashes?

I wasn't able to test it, do you know how to do it? If so, can you test that?